### PR TITLE
Verbosity=0 should mean silent

### PR DESCRIPTION
**Pull request by bcipolli**
Fix for #366

The `verbosity` parameter should control output from a function, with `0` meaning no output (and a non-zero default for functions that we want to output by default.  In some functions, current behavior was to always output. 

Here, I have fixed the above, tweaked docstrings, and changed places where `verbose` was passed as a boolean to be passed as an int.

I've also made changes to the testing code, to suppress output by forcing `verbose=0`.  With this, we can expect no output in tests--allowing browsing test results more cleanly and exposing more clearly unexpected prints and warnings.

### CODE REVIEW

**Comment by lesteve** [comment]
There is a lingering `self.verbose > 10` in nilearn/input_data/nifti_masker.py.

File: nilearn/datasets.py [hunk 1]
```diff
@@ -439,7 +455,8 @@ def _fetch_file(url, data_dir, resume=True, overwrite=False,
     try:
         # Download data
         displayed_url = urllib.splitquery(url)[0] if verbose == 0 else url
```
Comment (lesteve) [inline nilearn/datasets.py]: I guess this one should become `if verbose == 1` (shortening of the url for easier reading with the default verbosity)

File: nilearn/tests/test_datasets.py [hunk 2]
```diff
@@ -27,14 +27,22 @@
 file_mock = None
 
 
+def setup_silence():
+    # Wrap functions with versions that have verbose=0
+    datasets._get_dataset_dir = silence_fn(datasets._get_dataset_dir)
+    datasets._uncompress_file = silence_fn(datasets._uncompress_file)
+    datasets._fetch_file = silence_fn(datasets._fetch_file)
+    datasets._fetch_files = silence_fn(datasets._fetch_files)
+    datasets._chunk_read_ = silence_fn(datasets._chunk_read_)
```
Comment (lesteve) [inline nilearn/tests/test_datasets.py]: I am sligthly uneasy with this for two reasons:
* you are silencing all functions at once, which means that you could hide potential issues, for example if a verbose parameter wasn't passed from the caller to the callee (you fixed a few of those in this PR by the way, thanks for that)
* ideally, it'd be good to have a teardown equivalent of this that sets back the datasets functions to their original selves, so that each test doesn't modify the global state. I am guessing that with the current version of the code, you only need one decorated test to silence all of them.

Comment (lesteve) [inline nilearn/tests/test_datasets.py]: Actually thinking about it a bit more, why is it not enough to pass verbose=0 into the tested functions? Was it done to make it easier on the test writer so that he doesn't have to remember to add verbose=0 to each datasets function call?

Comment (GaelVaroquaux) [inline nilearn/tests/test_datasets.py]: > Actually thinking about it a bit more, why is it not enough to pass verbose=0
> into the tested functions?

That would seem like the right thing to do to me. Unless the purpose is
to trigger the printing code in order to test it.

Comment (bcipolli) [inline nilearn/tests/test_datasets.py]: It seems more failsafe to me.  To test the output directly, just don't add the decorator to the test... or the decorator could be extended to accept a list of function names to silence, or to avoid silencing.

I don't have any strong opinion about it, so happy to simply pass `verbose=0` explicitly, or to write teardown code.

Comment (bcipolli) [inline nilearn/tests/test_datasets.py]: On further thought, I think the *best* way to do this is to automatically silence all functions--there should be no manual verification in tests, so output simply isn't helpful.  To test verbosity of any function, people should write specific unit tests.

What I've done is shorthand for that.  I could use the `inspect` function to recursively search modules for functions that accept a `verbose` parameter, and I could write a teardown function as well.

This seems the most complete and safest to me.  It could also wrap `stderr` / `stdout` to test whether any function has output... in fact, doing this for the entire test suite via a `TestRunner`, rather than using a decorator, is also an option.

Comment (lesteve) [inline nilearn/tests/test_datasets.py]: tl; dr: it just seems to me that passing `verbose=0` into the needed functions is a lot simpler.

It does the job we want which is to reduce noise in the output of nosetests. Yes we may forget to set `verbose=0` occasionally when adding a new test, but we will probably notice eventually when either running the tests or looking at the travis-ci output.

Also as I mentioned above, it could be a way to potentially detect bugs, like the verbose parameters that aren't passed down the call stack. Let's be honest, I don't think we have so much time on our hands that we are going to unit test each of our print statements.

> To test the output directly, just don't add the decorator to the test

That's actually not true because you don't have a teardown method and you only need one decorated test with setup_silence to silence your dataset functions for all the tests that are going to run from then on. Changing the global state in tests is very very nasty.

An example of the side-effects, add this test at the end of test_datasets.py that checks that some printing on stdout actually occurs:

```python

import sys
from contextlib import contextmanager

@contextmanager
def redirect_stdout(new_target):
    old_target, sys.stdout = sys.stdout, new_target # replace sys.stdout
    try:
        yield new_target # run some code with the replaced stdout
    finally:
        sys.stdout = old_target # restore to the previous value

def test_stdout_miyawaki2008():
    with redirect_stdout(open('/tmp/test', 'w')):
        datasets._get_dataset_dir('test', verbose=5)

    content = open('/tmp/test').read()
    assert_true('search paths' in content)

```

This won't pass with you current code but it will pass if you remove your setup_silence decorators (or have setup_silence be just a no-op)

**Comment by bcipolli** [comment]
OK, pulled the `silence_fn` and added `verbosity=0` to all relevant tests in `test_datasets.py`.  Made the two verbosity tweaks above as well, did a little commit squashing... I think this is all set!

**Comment by bcipolli** [comment]
...and with that said, the output is back in the tests.  Investigating.

**Comment by bcipolli** [comment]
After changing how the testing was done, then searching through the output again, I found quite a few more changes to make things work as advertised.  Please review the last 7 commits carefully; the changes are small but took thinking.

It would also be good to have a test run with all tests using `verbose=10`, to make sure none of the printing code has errors in it.  I suggest to add an issue for this.

File: nilearn/input_data/multi_nifti_masker.py [hunk 3]
```diff
@@ -185,7 +185,7 @@ def fit(self, imgs=None, y=None):
                             target_shape=self.target_shape,
                             n_jobs=self.n_jobs,
                             memory=self.memory,
-                            verbose=(self.verbose - 1),
+                            verbose=max(0, self.verbose - 1),
```
Comment (lesteve) [inline nilearn/input_data/multi_nifti_masker.py]: There are other occurences of this verbose - 1 elsewhere in the code. Would you mind fixing them while you are at it?

```
~/dev/nilearn $ git grep -n -P 'verbose\s*-\s*1'
nilearn/group_sparse_covariance.py:548:                verbose=self.verbose - 1, debug=False)
nilearn/input_data/multi_nifti_masker.py:188:                            verbose=(self.verbose - 1),
nilearn/input_data/nifti_masker.py:168:                verbose=(self.verbose - 1),
```

Comment (lesteve) [inline nilearn/input_data/multi_nifti_masker.py]: Actually you have already, thanks for that and sorry for the noise ...

File: nilearn/input_data/nifti_masker.py [hunk 4]
```diff
@@ -186,7 +186,7 @@ def fit(self, imgs=None, y=None):
             self.affine_ = self.mask_img_.get_affine()
         # Load data in memory
         self.mask_img_.get_data()
-        if self.verbose > 10:
+        if self.verbose > 0:
```
Comment (lesteve) [inline nilearn/input_data/nifti_masker.py]: I am not too sure whether > 10 was a typo or meant to indicate a super high verbosity was needed. At least the required verbosity is the same as the other print statements in the same class so that's probably fine.

Comment (GaelVaroquaux) [inline nilearn/input_data/nifti_masker.py]: Indeed, I think that it should be a high value. Printing just before a return is not someone wants often: most often we know that the operation is finished because either another one is starting, or the processing has finished.

**Comment by lesteve** [comment]
Sorry there is a tiny merge conflict I introduced by fixing something completely unrelated in master. Do you mind rebasing on master ? You just need to rename files -> filenames in `nilearn/_utils/testing.py`:

```
<<<<<<< HEAD
        kwargs['mock'] = True
        filenames = original_fetch_files(*args, **kwargs)
=======
        files = self._mock_fetch_files(*args, **kwargs)
>>>>>>> 33c9bdbb9bbafc437af606621ef6b269af9462a3
```

**Comment by lesteve** [comment]
LGTM, other than the verbose level in nilearn/input_data/nifti_masker.py and the merge conflict.

Maybe @AlexandreAbraham you can have a quick look ?

**Comment by AlexandreAbraham** [comment]
I agree with you!

**Comment by bcipolli** [comment]
OK, I rebased to master and restored the one verbosity check to 10 where suggested.  The tests pass and the output still looks clean, so this should be good to go!

**Comment by lesteve** [comment]
Thanks a lot for this, clearly a nice improvement, be it only to make travis logs load a lot faster.
